### DISCUSSION TITLE
How to limit matches to "exact" results

### DISCUSSION THREAD

**Discussion by keith** [ast-grep/ast-grep#576] [Q&A]
I have this `sg` command: `sg -p 'override func setUp() { super.setUp() }' -l swift`. I would expect this to only match:

```swift
override func setUp() {
    super.setUp()
}
```

and not:

```swift
override func setUp() {
    super.setUp()
    print("more statements")
}
```

But it does actually match both. Is there special syntax to restrict the match somehow? Otherwise is this type of issue up to `sg` or the Swift specific tree-sitter parser?

**Comment by HerringtonDarkholme** [comment]
It looks like that the behavior is tree-sitter-swift specific. 

Other languages, say [JavaScript](https://ast-grep.github.io/playground.html#eyJtb2RlIjoiUGF0Y2giLCJsYW5nIjoiamF2YXNjcmlwdCIsInF1ZXJ5IjoiZnVuY3Rpb24gdGVzdCgpIHtcbiAgICB0ZXN0KClcbn0iLCJyZXdyaXRlIjoibG9nZ2VyLmxvZygkTUFUQ0gpIiwiY29uZmlnIjoiIyBZQU1MIFJ1bGUgaXMgbW9yZSBwb3dlcmZ1bCFcbiMgaHR0cHM6Ly9hc3QtZ3JlcC5naXRodWIuaW8vZ3VpZGUvcnVsZS1jb25maWcuaHRtbCNydWxlXG5ydWxlOlxuICBhbnk6XG4gICAgLSBwYXR0ZXJuOiBjb25zb2xlLmxvZygkQSlcbiAgICAtIHBhdHRlcm46IGNvbnNvbGUuZGVidWcoJEEpXG5maXg6XG4gIGxvZ2dlci5sb2coJEEpIiwic291cmNlIjoiZnVuY3Rpb24gdGVzdCgpIHtcbiAgdGVzdCgpXG4gIHByaW50KFwibW9yZVwiKVxufVxuZnVuY3Rpb24gdGVzdCgpIHtcbiAgdGVzdCgpXG59XG4ifQ==) does not have the issue at all.

Unfortunately, we don't have swift on the playground yet. And it is not easy to debug tree-sitter with CLI only.

I'm planning to add the playground support recently. And then I can vet the issue here.

**Reply by HerringtonDarkholme** [reply]
I have added [Swift AST Playground](https://ast-grep.github.io/playground.html#eyJtb2RlIjoiUGF0Y2giLCJsYW5nIjoic3dpZnQiLCJxdWVyeSI6ImZ1bmMgYSgpIHtcbiAgICBiXG59IiwicmV3cml0ZSI6ImxvZ2dlci5sb2coJE1BVENIKSIsImNvbmZpZyI6IiMgWUFNTCBSdWxlIGlzIG1vcmUgcG93ZXJmdWwhXG4jIGh0dHBzOi8vYXN0LWdyZXAuZ2l0aHViLmlvL2d1aWRlL3J1bGUtY29uZmlnLmh0bWwjcnVsZVxucnVsZTpcbiAgYW55OlxuICAgIC0gcGF0dGVybjogY29uc29sZS5sb2coJEEpXG4gICAgLSBwYXR0ZXJuOiBjb25zb2xlLmRlYnVnKCRBKVxuZml4OlxuICBsb2dnZXIubG9nKCRBKSIsInNvdXJjZSI6ImZ1bmMgYSgpIHsgXG4gIGJcbiAgY1xufSJ9)

The issue here is the `statements` AST.

|two statemtements| one statement|
|--|--|
|<img width="342" alt="image" src="https://github.com/ast-grep/ast-grep/assets/2883231/bebb7c5e-db8f-455e-bb4f-da367f821234">|<img width="298" alt="image" src="https://github.com/ast-grep/ast-grep/assets/2883231/fe54de91-5254-4aed-940b-63bf26f66cd0">|

ast-grep will first try matching `function_body`. And then see body's direct children `{`, `statements`, and `}` matches.
Finally, two `statements` matches because the two-statement source contains all statement from query.

Let me think about how the matching algorithm should work here....

**Reply by php-coder** [reply]
Any updates on this? :)

**Reply by HerringtonDarkholme** [reply]
I don't think this is an easy fix. You can use YAML `not` rule to filter unwanted matching instead, at the moment.
